CustomShips.cpp

```diff
@@ -931,6 +932,70 @@ HOOK_METHOD(ShipManager, DamageSystem, (int roomId, Damage dmg) -> void)
     super(roomId, dmg);
 }
 
+HOOK_METHOD(ShipManager, DamageArea, (Pointf location, Damage dmg, bool forceHit) -> bool)
+{
+    LOG_HOOK("HOOK_METHOD -> ShipManager::DamageArea -> Begin (CustomDamage.cpp)\n")
+
+    int roomId = ship.GetSelectedRoomId(location.x, location.y, true);
+    auto ex = RM_EX(ship.vRoomList[roomId]);
+    bool resist = false;
+
+    if (random32() % 100 < ex->hullDamageResistChance && dmg.iDamage > 0)
+    {
+        dmg.iSystemDamage = +dmg.iDamage;
+        dmg.iPersDamage = +dmg.iDamage;
+        dmg.iDamage = 0;
+        resist = true;
+    }
+
+    bool ret = super(location, dmg, forceHit);
+
+    if (resist && ret)
+    {
+        auto msg = new DamageMessage(1.f, ship.GetRoomCenter(roomId), DamageMessage::MessageType::RESIST);
+        msg->color.r = 127.f / 255.f;
+        msg->color.g = 127.f / 255.f;
+        msg->color.b = 127.f / 255.f;
+        msg->color.a = 1.f;
+        damMessages.push_back(msg);
+    }
+    
+    return ret;
+}
+
+HOOK_METHOD(ShipManager, DamageBeam, (Pointf location1, Pointf location2, Damage dmg) -> void)
+{
+    LOG_HOOK("HOOK_METHOD -> ShipManager::DamageBeam -> Begin (CustomDamage.cpp)\n")
+
+    int room1 = ship.GetSelectedRoomId(location1.x, location1.y, true);
+    int room2 = ship.GetSelectedRoomId(location2.x, location2.y, true);
+    bool willDamage = (room1 != room2);
+    if (willDamage)
+    {
+        auto ex = RM_EX(ship.vRoomList[willDamage]);
```

Comment:
Using the variable willDamage as the room index will always return you 1, meanning that only the `hullDamageResistChance`  of Room 1 will be taken into account for a damaging beam hull resist.

You have to dynamically take into account which room is being damaged at the given tick (it always start with room1, then room2 for the rest of the beam damage tick), this condition takes care of that.
```suggestion
    if (room1 != room2)
    {
        auto ex = room2 != -1 ? RM_EX(ship.vRoomList[room2]) : RM_EX(ship.vRoomList[room1]);
```